README.md
# Domogik Plugin plcbus

## Purpose

This is a package for Domogik : http://www.domogik.org

Domogik is an open source home automation solution.

## Documentation 

You can find the documentation source in the **docs/** folder. When the package will be installed, the documentation will be available in the **Documentation** menu of the Domogik administration for this package.
You may also find online documentation for this plugin. You will be able to find the documentation url on http://repo-public.domogik.org/dashboard

## Install the package

To install this package on your Domogik system, you can go in this GitHub repository releases page and get the link to a release .zip file. Then you just have to do :

    dmg_package -i http://path.to/the/file.zip


docs/changelog.txt
=========
Changelog
=========

0.1
===

* Plugin creation
* Adapt the old domogik 0.3 from : Yoann HINARD <yoann.hinard at gmail.com>, Francois PINET <domopyx at gmail.com>


tests/001_configure.py
#!/usr/bin/python
#-*- coding: utf-8 -*-

from domogik.tests.common.helpers import configure, delete_configuration
from domogik.common.utils import get_sanitized_hostname

plugin =  "plcbus"

host_id = get_sanitized_hostname()
delete_configuration("plugin", plugin, host_id)

configure("plugin", plugin, host_id, "device", "/dev/plcbus")
configure("plugin", plugin, host_id, "usercode", "D1")
configure("plugin", plugin, host_id, "probe-list", "A")
configure("plugin", plugin, host_id, "usercode", "D1")
configure("plugin", plugin, host_id, "probe-interval", 60)
configure("plugin", plugin, host_id, "configured", True)


tests/validate_json.py
#!/usr/bin/python
#-*- coding: utf-8 -*-

### configuration ###########################################
plugin = "plcbus"
########################################################

import json
import sys
import os
from domogik.common.packagejson import PackageJson

json_file = "/var/lib/domogik/domogik_packages/plugin_{0}/info.json".format(plugin)
if os.path.exists(json_file) :
    json_file = os.path.realpath(json_file)
    data = json.load(open(json_file))
    print data
    print "Package validation ..."
    p = PackageJson(path=json_file)
    if p.validate() == False :
        print "*** Error json validation :{0}".format(json_file)
    else :
        print "Json is validate :{0}".format(json_file)
else :
    print "*** Error *** Don't find file :{0}".format(json_file)


tests/003_deldevices.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from domogik.tests.common.testdevice import TestDevice
from domogik.common.utils import get_sanitized_hostname
import traceback
import sys

if __name__ == "__main__":

    # set up the plugin name
    name = "plcbus"

    # load the test devices class
    td = TestDevice()

    # delete existing devices for this plugin on this host
    client_id = "{0}-{1}.{2}".format("plugin", name, get_sanitized_hostname())
    try:
        td.del_devices_by_client(client_id)
    except:
        print(u"Error while deleting all the test device for the client id '{0}' : {1}".format(client_id, traceback.format_exc()))
        sys.exit(1)


lib/plcbus.py
# -*- coding: utf-8 -*-

""" This file is part of B{Domogik} project (U{http://www.domogik.org}).

License
=======

B{Domogik} is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

B{Domogik} is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Domogik. If not, see U{http://www.gnu.org/licenses}.

Plugin purpose
==============

Support PLCBUS power-based technology

Implements
==========

- PLCBUSException:.def __init__(self, value):
- PLCBUSException:.def __str__(self):
- PLCBUSAPI:.def __init__(self, serial_port_no):
- PLCBUSAPI:.def _valid_item(self, item):
- PLCBUSAPI:.def _valid_house(self, house):
- PLCBUSAPI:.def _valid_usercode(self, item):
- PLCBUSAPI:.def _convert_device_to_hex(self, item):
- PLCBUSAPI:.def _convert_data(self, data):
- PLCBUSAPI:.def send(self, cmd, item, ucod, data1, data2):
- PLCBUSAPI:.def get_all_on_id(self, usercode, housecode):

@author: Domogik project
@copyright: (C) 2007-2016 Domogik project
@license: GPL(v3)
@organization: Domogik
"""

from domogik_packages.plugin_plcbus.lib.PLCBusSerialHandler import serialHandler


class PLCBUSException(Exception):
    '''
    PLCBUS Exception
    '''

    def __init__(self, value):
        Exception.__init__(self)
        self.value = value

    def __str__(self):
        return repr(self.value)


class PLCBUSAPI:
    '''
    This class define some facilities to use PLCBUS.
    ALL_USER_UNIT_OFF must be with home unit=00.
    '''

    def __init__(self, log, serial_port_no, command_cb, message_cb):
        """ Main PLCBus manager
        Use serialHandler for low-level serial management
        @param log : log instance
        @param serial_port_no : Number or path of the serial port
        @param command_cb: callback called when a command has been succesfully sent
        @param message_cb: called when a message is received from somewhere else on the network
        """
        self._log = log
        #For these 2 callbacks, the param is sent as an array
        self._housecodes = list('ABCDEFGHIJKLMNOP')
        self._valuecode = enumerate(self._housecodes)
        self._codevalue = dict([(v, k) for (k, v) in self._valuecode])
        self._unitcodes = range(1, 17)
        self._usercodes = list('0123456789ABCDEF')
        self._cmdplcbus = {
            #020645000800000c03 because the remote send all user unit off
            'ALL_UNITS_OFF': '00',
            'ALL_LIGHTS_ON': '01',
            'ON': '22', #ON and ask to send ACK (instead of '02')
            'OFF': '23', #OFF and send ACK
            'DIM': '24',
            'BRIGHT': '25',
            'ALL_LIGHTS_OFF': '06',
            'ALL_USER_LTS_ON': '07',
            'ALL_USER_UNIT_OFF': '08',
            'ALL_USER_LIGHT_OFF': '09',
            'BLINK': '2a',
            'FADE_STOP': '2b',
            'PRESET_DIM': '2c',
            'STATUS_ON': '0d',
            'STATUS_OFF': '0e',
            'STATUS_REQUEST': '0f',
            'REC_MASTER_ADD_SETUP': '30',
            'TRA_MASTER_ADD_SETUP': '31',
            'SCENE_ADR_SETUP': '12',
            'SCENE_ADR_ERASE': '13',
            'ALL_SCENES_ADD_ERASE': '34',
            #'FOR FUTURE': '15',
            #'FOR FUTURE': '16',
            #'FOR FUTURE': '17',
            'GET_SIGNAL_STRENGTH': '18',
            'GET_NOISE_STRENGTH': '19',
            'REPORT_SIGNAL_STREN': '1a',
            'REPORT_NOISE_STREN': '1b',
            'GET_ALL_ID_PULSE': '1c',
            'GET_ALL_ON_ID_PULSE': '1d',
            'REPORT_ALL_ID_PULSE': '1e',
            'REPORT_ONLY_ON_PULSE': '1f'}
        #instead of using serial directly, use serialHandler
        self._ser_handler = serialHandler(serial_port_no, command_cb, message_cb)
        self._ser_handler.start()

    def _valid_item(self, item):
        '''
        Check an item to have good 'HU' syntax
        Raise exception if it is not
        '''
        h, u = (item[0].upper(), item[1])
        try:
            if not (h in self._housecodes and int(u) in self._unitcodes):
                raise AttributeError
        except:
            self._log.warning("Invalid item %s%s, must be 'HU'" % (h, u))

    def _valid_house(self, house):
        '''
        Check an house to have good 'H' syntax
        Raise exception if it is not
        '''
        try:
            if house[0] not in self._housecodes:
                raise AttributeError
        except:
            self._log.warning("Invalid house %s, must be 'H' format, between A and P" % house[0].upper())

    def _valid_usercode(self, item):
        '''
        Check an user code to have good 'H' syntax
        Raise exception if it is not
        '''
        h, u = (item[0].upper(), item[1])
        try:
            if not (h in self._usercodes and int(u) in self._usercodes):
                raise AttributeError
        except:
            self._log.warning("Invalid user code %s, must be 'H' format, between 00 and FF" % h)

    def _convert_device_to_hex(self, item):
        if item == None or len(item) == 0:
            return "00"
        elif len(item) == 1:
            return '%01x0' % (self._codevalue[item[0]])
        else: 
            var1 = int(item[1:]) - 1
            var2 = '%01X%01x' % (self._codevalue[item[0]], var1)
            return var2

    def _convert_data(self, data):
        # result must have 2 caracters
        var1 = hex(int(data))[2:]
        if len(var1) == 2:
            var2 = '%01X' % (int(data))
            return var2
        else:
            var2 = '0%01X' % (int(data))
            return var2

    def send(self, cmd, item, ucod, data1 = "00", data2 = "00"):
        # after cmd add level, rate : put in data1 and data2
        # (just data1 for these cases)
        '''
        Send a command PLCBUS to 1141 plugin
        @param cmd : Command to send ('ON','OFF', etc)
        @param item : Item to send order to (must be "HU" format)
        @param ucod : User code of item (must be 'H' syntax between 00 to FF)
        '''
        # TODO : test on ALL_USER_UNIT_OFF
        #try:
        try:
            command = self._cmdplcbus[cmd]
        except KeyError:
            print("PLCBUS Frame generation error, command does not exist ", cmd)
        else:
            if cmd == 'ALL_UNITS_OFF':
                plcbus_frame = '020645000800000c03'
            else:
                plcbus_frame = '0205%s%s%s%s%s03' % (ucod,
                    self._convert_device_to_hex(item), self._cmdplcbus[cmd],
                    self._convert_data(data1), self._convert_data(data2))
            try:
                message = plcbus_frame.decode('HEX')
            except TypeError:
                print("PLCBUS Frame generation error, does not result in a HEX string ", plcbus_frame)
            else:
                self._ser_handler.add_to_send_queue(plcbus_frame)

    def get_all_on_id(self, usercode, housecode):
        '''
        Fastpoll the housecode and return every on plugins
        @param usercode : User code of item (must be 'H' syntax between 00 to
        FF)
        @param housecode : one or more housecodes
        '''
        onlist = []
        self.send("GET_ALL_ON_ID_PULSE", housecode + "1", usercode)
        response = self._ser_handler.get_from_answer_queue()
        if(response):
            print "Hoora response received", response
            data = int(response[10:14], 16)
            for i in range(0, 16):
                if data >> i & 1:
                    onlist.append(housecode + str(self._unitcodes[i]))
        print "on :", onlist
        return onlist

    def stop(self):
        """ Ask thread to stop
        """
        self._log.debug("Stopping plcbus serial library")
        self._ser_handler.stop()


lib/PLCBusSerialHandler.py
# -*- coding: utf-8 -*-

""" This file is part of B{Domogik} project (U{http://www.domogik.org}).

License
=======

B{Domogik} is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

B{Domogik} is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Domogik. If not, see U{http://www.gnu.org/licenses}.

Plugin purpose
==============

Get event from PLCBUS and send them on Xpl

Implements
==========

- serialHandler

@author: Yoann HINARD <yoann.hinard@gmail.com>
@copyright: (C) 2007-2016 Domogik project
@license: GPL(v3)
@organization: Domogik
"""

import sys
import time
from binascii import hexlify
import Queue
import threading
import mutex
import datetime
import serial



class serialHandler(threading.Thread):
    """
    Threaded class to handle serial port in PLCbus communication
    Send PLCBUS frames when available in the send_queue and manage
    retransmission if needed
    Put received frames in the receveive_queue (to be sent on the MQ network
    """

    def __init__(self, serial_port_no, command_cb, message_cb):
        """ Initialize threaded PLCBUS manager
        Will handle communication to and from PLCBus 
        @param serial_port_no : Number or path of the serial port 
        @param command_cb: callback called when a command has been succesfully sent
        @param message_cb: called when a message is received from somewhere else on the network
        For these 2 callbacks, the param is sent as an array
        """
        threading.Thread.__init__(self)
        #invoke constructor of parent class
        self._ack = threading.Event() #Shared list between reader and writer
        self._need_answer = ["1D", "1C"]
        self._stop = threading.Event()
        self._has_message_to_send = threading.Event()
        #serial port init
        self.__myser = serial.Serial(serial_port_no, 9600, timeout = 0.4,
                parity = serial.PARITY_NONE, stopbits=serial.STOPBITS_ONE,
                xonxoff = 0) #, rtscts=1)
#        self._want_lock = threading.Event()
#        self._mutex = mutex.mutex()
        self._send_queue = Queue.Queue()
        self._cb = command_cb
        self._msg_cb = message_cb
        #self._reader = self.__Reader(self.__myser, self._want_lock, self._mutex, self._ack, message_cb)
        #self._reader.start()
        #self._writer = self.__Writer(self.__myser, self._want_lock, self._mutex, self._ack, command_cb, self._reader)
        #self._writer.start()

    def _send(self, plcbus_frame):
        #Resend if proper ACK not received
        #check for ack pulse
        explicit_frame = self.explicit_message(plcbus_frame)
        if (int(plcbus_frame[8:10], 16) >> 5) & 1: #ACK pulse bit set to 1
            #The ACK message take only 10ms + 10ms to bring it back to the computer.
            #Anyway, it seems that the mean time before reading the ack is about 
            #0.6s . Because there is another sleep in the loop, this sleep is only 0.3s
         #   time.sleep(0.3)
            ACK_received = 0
            # like a timer, does not wait for more than 2seconds for example
            time1 = time.time()
            while not self._stop.isSet():
                #The ack message is sent immediately after the message has been received 
                #and transmission time is 20mS (10ms for message propagation to adapter,
                #and 10ms between adapter and computer
                #We sleep 20ms between each check
                self.needs_ack_for(plcbus_frame)
                self._basic_write(plcbus_frame)
                time.sleep(0.6)
                print("time before wait : %s" % time.time())
                for i in range(3):
                    self.receive()
                    if self._ack.isSet():
                        print("got ack in first read")
                        break
                print("time after wait : %s" % time.time())
                if self._ack.isSet():
                    print(plcbus_frame + " : Ack set")
                    ACK_received = 1
                    self._ack.clear()
                    self._cb(self.explicit_message(plcbus_frame))
                    break

                if (ACK_received == 1):
                    break
                elif(time1 + 3.1 < time.time()):
                    print("WARN : Message %s sent, but ack never received" % plcbus_frame)
                    break #2s
        elif explicit_frame["d_command"] not in ['GET_ALL_ID_PULSE', 'GET_ALL_ON_ID_PULSE']:
            #No ACK asked, we consider that the message has been correctly sent
            self._basic_write(plcbus_frame)
            self._cb(explicit_frame)
        else:
            self._basic_write(plcbus_frame)

    def _basic_write(self, frame):
        """Write a frame on serial port
        This method should only be called as mutex.lock() parameter
        @param frame : The frame to write 
        """
        print("SEND : %s" % frame)
        self.__myser.write(frame.decode("HEX"))

    def add_to_send_queue(self, trame):
        print("add_to_send_queue : %s" % trame)
        self._send_queue.put(trame)

    def needs_ack_for(self, frame):
        """ Called by the Writer to let Reader knows that a ACK is waited
        it will set the internal _waited_ack until the ack is received 
        @param frame : the plcbus frame
        """
        self._waited_ack = frame

    def _is_ack_for_message(self, m1, m2):
        #check the ACK bit
        #print("ACK check " + m1.encode('HEX') +" " + m2)
        #check house code and user code in hexa string format like '45E0'
        if(m1[4:8].upper() == m2[4:8].upper()) and (m1[8:10].upper() == m2[8:10].upper()): #Compare user code + home unit
            #print("housecode and usercode OK")
            return (int(m1[14:16], 16) & 0x20) #test only one bit
        return False

    def _is_answer(self, message):
        # if command is in answer required list (not ACK required, it's
        # different)
        # if R_ID_SW bit set
        # maybe pass this list to the _init_ of this handler to make it
        # compatible with other protocols
        if((int(message[14:15], 16) >> 2 & 1) and message[8:10].upper() in
                ["1C","1D"]):
            return True
        return False

    def explicit_message(self, message):
        """ Parse a frame 
        """
        cmdplcbus = {
        '00': 'ALL_UNITS_OFF',
        '01': 'ALL_LIGHTS_ON',
        '22': 'ON', #ON and ask to send ACK (instead of '02')
        '23': 'OFF', #OFF and send ACK
        '24': 'DIM',
        '25': 'BRIGHT',
        '06': 'ALL_LIGHTS_OFF',
        '07': 'ALL_USER_LTS_ON',
        '08': 'ALL_USER_UNIT_OFF',
        '09': 'ALL_USER_LIGHT_OFF',
        '2a': 'BLINK',
        '2b': 'FADE_STOP',
        '2c': 'PRESET_DIM',
        '0d': 'STATUS_ON',
        '0e': 'STATUS_OFF',
        '0f': 'STATUS_REQUEST',
        '30': 'REC_MASTER_ADD_SETUP',
        '31': 'TRA_MASTER_ADD_SETUP',
        '12': 'SCENE_ADR_SETUP',
        '13': 'SCENE_ADR_ERASE',
        '34': 'ALL_SCENES_ADD_ERASE',
        '15': 'FOR FUTURE',
        '16': 'FOR FUTURE',
        '17': 'FOR FUTURE',
        '18': 'GET_SIGNAL_STRENGTH',
        '19': 'GET_NOISE_STRENGTH',
        '1a': 'REPORT_SIGNAL_STREN',
        '1b': 'REPORT_NOISE_STREN',
        '1c': 'GET_ALL_ID_PULSE',
        '1d': 'GET_ALL_ON_ID_PULSE',
        '1e': 'REPORT_ALL_ID_PULSE',
        '1f': 'REPORT_ONLY_ON_PULSE'}
        home = "ABCDEFGHIJKLMNOP"
        r = {}
        r["start_bit"] = message[0:2]
        r["data_length"] = int(message[2:4])
        int_length = int(message[2:4])*2
        r["data"] = message[4:4+int_length]
        r["d_user_code"] = r["data"][0:2]
        r["d_home_unit"] = "%s%s" % (home[int(r["data"][2:3], 16)],int(r["data"][3:4], 16)+1)
        r["d_command"] = cmdplcbus[r["data"][4:6]]
        r["d_data1"] = int(r["data"][6:8],16)
        r["d_data2"] = int(r["data"][8:10],16)
        if r["data_length"] == 6:
            r["rx_tw_switch"] = r["data"][11:]
        r["end_bit"] = message[-2:]
        return r

    def receive(self):
        #Avoid to wait if there is nothing to read
#        try:
        if self._stop.isSet():
            return
        if self.__myser.inWaiting() < 9:
            time.sleep(0.4)
            return
        while self.__myser.inWaiting() >= 9:
            message = self.__myser.read(9) #wait for max 400ms if nothing to read
#        except IOError:
#            pass
            if(message):
                m_string = hexlify(message)
                #self.explicit_message(m_string)
                #if message is likely to be an answer, put it in the right queue
                #First we check that the message is not from the adapter itself
                #And simply ignore it if it's the case 
                print("str : %s" % m_string)
                if self._is_from_myself(m_string):
                    print("from myself")
                    return
                if self._is_answer(m_string):
                    print("ANSWER : %s" % m_string)
                    self._cb(self.explicit_message(m_string))
                elif self._is_ack(m_string):
                    print("IS ACK : %s, waited ack : %s" % (m_string, self._waited_ack))
                    if (self._waited_ack != None) and self._is_ack_for_message(m_string, self._waited_ack):
                        self._waited_ack = None
                        self._ack.set()
                else:
                    print("QUEUE : %s" % m_string)
                    self._cb(self.explicit_message(m_string))

    def stop(self):
        """ Ask the thread to stop, 
        will only set a threading.Event instance
        and close serial port
        """
        self._stop.set()
        self.__myser.close()

    def run(self):
        #serial handler main thread
#        self._mutex.testandset()
        while not self._stop.isSet():
            #The Event _has_message_to_send is only used to optimize
            #The test isSet is much faster than the empty() test 
            #If _has_message_to_send is locked, then there is at least 1 lock(function)
            #in the queue, so the unlock() will just do this call, 
            #not really unlock the mutex.
            try:
                item = self._send_queue.get_nowait()
            except Queue.Empty:
                pass
            else:
                self._send(item)
#                self._mutex.unlock()
#                self._mutex.testandset()
            #print("receiving")
            self.receive()

    def _is_ack(self, message):
        """ Check if a message is an ack 
            @param message : message to check
        """
        return int(message[14:16], 16) & 0x20

    def _is_from_myself(self, message):
        """ Check if a message is sent by the adapter itself
            @param message : message to check
        """
        return int(message[14:16], 16) & 0x10

#a = serialHandler()
#a.start()
#pas sur du contenu des trames suivantes
#trame = '0205000000000003'
#trame = '0205000102000003'#A2 on
#trame = '0205000002000003'#A1 on
#trame = '0205450122000003' #A2 on ack asked
#trame = '020545E302000003' #B1 on
#a.add_to_send_queue(trame)

#a.get_from_receive_queue() #attention, bloquant


#je n'ai pas gere comment quitter
#a.join()



#trame = '0205FF0123000003' #A2 off ack asked
#trame = '020500000F000003' #Status checking
#trame = '0205000018000003' #get signal strength
